test/Bound.lean

```diff
@@ -0,0 +1,138 @@
+lemma test_coe (x y : ℝ≥0) (h : x < y) : (x : ℝ) < y := by bound
+lemma test_dist : dist a c ≤ dist a b + dist b c := by bound
+lemma test_log (x y : ℝ) (x0 : 0 < x) (h : x ≤ y) : x.log ≤ y.log := by bound
+end bound_tests
+
+section guess_tests
+variable {a b c : ℝ}
+variable {n m : ℕ}
+lemma test_le_max_of_le_left (h : a ≤ b) : a ≤ max b c := by bound
+lemma test_le_max_of_le_right (h : a ≤ c) : a ≤ max b c := by bound
+lemma test_lt_max_of_lt_left (h : a < b) : a < max b c := by bound
+lemma test_lt_max_of_lt_right (h : a < c) : a < max b c := by bound
+lemma test_min_le_of_left_le (h : a ≤ c) : min a b ≤ c := by bound
+lemma test_min_le_of_right_le (h : b ≤ c) : min a b ≤ c := by bound
+lemma test_min_lt_of_left_lt (h : a < c) : min a b < c := by bound
+lemma test_min_lt_of_right_lt (h : b < c) : min a b < c := by bound
+lemma test_pow_le_pow_right (a1 : 1 ≤ a) (h : m ≤ n) : a^m ≤ a^n := by bound
+lemma test_pow_le_pow_of_le_one (a0 : 0 ≤ a) (a1 : a ≤ 1) (h : n ≤ m) : a^m ≤ a^n := by bound
+lemma test_rpow_le_rpow_of_exponent_le (a1 : 1 ≤ a) (h : b ≤ c) : a^b ≤ a^c := by bound
+lemma test_rpow_le_rpow_of_exponent_ge (a0 : 0 < a) (a1 : a ≤ 1) (h : c ≤ b) : a^b ≤ a^c := by bound
+end guess_tests
+
+-- This breaks without appropriate g.withContext use
+lemma test_with_context {s : Set ℂ} (o : IsOpen s) (z) (h : z ∈ s) : ∃ r : ℝ, r > 0 := by
+  rw [Metric.isOpen_iff] at o
+  rcases o z h with ⟨t, tp, bs⟩
+  exists t/2
+  clear o h bs z s
+  bound
+
+-- Test various elaboration issues
+lemma test_try_elab {f : ℂ → ℂ} {z w : ℂ} {s r c e : ℝ}
+      (sc : ∀ {w}, abs (w - z) < s → abs (f w - f z) < e) (wz : abs (w - z) < s) (wr : abs w < r)
+      (h : ∀ z : ℂ, abs z < r → abs (f z) ≤ c * abs z) :
+      abs (f z) ≤ c * abs w + e := by
+  calc abs (f z) = abs (f w - (f w - f z)) := by ring_nf
+    _ ≤ abs (f w) + abs (f w - f z) := by bound
+    _ ≤ c * abs w + e := by bound [h w wr, sc wz]
+
+-- Test a lemma that requires function inference (and thus )
+lemma test_fun_inference {α : Type} {s : Finset α} {f g : α → ℂ} :
+    ‖s.sum (fun x ↦ f x + g x)‖ ≤ s.sum (fun x ↦ ‖f x + g x‖) := by
+  bound
+
+-- A test that requires reduction to weak head normal form to work (surfaced by `Hartogs.lean`)
+lemma test_whnf (x y : ℝ) (h : x < y ∧ True) : x ≤ y := by
+  bound [h.1]
+
+-- Used to fail with `unknown identifier n`, since I wasn't elaborating [] inside the goal
+theorem test_unknown_identifier {f : ℕ → ℝ} (le : ∀ n, f n ≤ n) : ∀ n : ℕ, f n ≤ n := by
+  intro n; bound [le n]
+
+-- Calc example: A weak lower bound for `z ← z^2 + c`
+lemma le_sqr_add {c z : ℂ} (cz : abs c ≤ abs z) (z3 : 3 ≤ abs z) :
+    2 * abs z ≤ Complex.abs (z^2 + c) := by
+  calc Complex.abs (z^2 + c)
+    _ ≥ Complex.abs (z^2) - abs c := by bound
+    _ ≥ Complex.abs (z^2) - abs z := by bound
+    _ ≥ (abs z - 1) * abs z := by rw [mul_comm, mul_sub_one, ←pow_two, ←Complex.abs.map_pow]
+    _ ≥ 2 * abs z := by bound
```

Comment:
Same comments as in the documentation

Reply:
Sorry about the `{Complex.,}abs` thing.  Unfortunately some of them work without it and some of them don't, due to imperfect inference, but I agree uniformity is good.